docs: supporting after for providers (#73743)
vercel/next.js: docs/01-app/02-building-your-application/10-deploying/index.mdx

</AppOnly>

<AppOnly>

### Using `after`

#### Self-hosting

`after` is fully supported in `next start`.

> Note that when you want to stop the server, you should give it a chance to gracefully shutdown by sending `SIGINT`/`SIGTERM` and waiting.

Reviewer: ```suggestion
Note that when you want to stop the server, you should allow it to gracefully shutdown by sending `SIGINT`/`SIGTERM` and waiting.
```

> Note that when you want to stop the server, you should give it a chance to gracefully shutdown by sending `SIGINT`/`SIGTERM` and waiting.
> This will allow the server to wait for any pending callbacks/promises passed to `after` to finish.

Reviewer: ```suggestion
  Note that when you want to stop the server, you should allow it to gracefully shutdown by sending `SIGINT`/`SIGTERM` and waiting.This will allow the server to wait for any pending callbacks/promises passed to `after` to finish.
```

#### Serverless hosting

> `after` is fully supported on Vercel. If you're using another serverless deployment platform, `after` may or may not work depending on platform support.

Reviewer: How's this? 

```suggestion
`after` is fully supported on Vercel. If you're using another serverless deployment platform, check if your provider supports `after`. 

> **For platform providers**: see the [supporting `after` section](#for-providers-supporting-after-on-your-serverless-platform) for more information.
```

> 

Author: wonky, needs cleanup imo

```

</PagesOnly>

<AppOnly>

## For providers: supporting `after` on your serverless platform

If you’re a hosting provider and want your users to be able to use `after` on your platform, this section applies to you.

### Background

> Using `after` in a serverless context requires a way to wait for asynchronous tasks to finish after the response has been sent.
> We achieve this using a primitive called `waitUntil(promise)`, which extends the lifetime of a serverless invocation
> until all promises passed to `waitUntil` have settled.
> On Vercel, this is done via [https://vercel.com/docs/functions/functions-api-reference#waituntil](https://vercel.com/docs/functions/functions-api-reference#waituntil).

Reviewer: ```suggestion
Using `after` in a serverless context requires waiting for asynchronous tasks to finish after the response has been sent. In Next.js and Vercel, this is achieved using a primitive called `waitUntil(promise)`, which extends the lifetime of a serverless invocation until all promises passed to [`waitUntil`](https://vercel.com/docs/functions/functions-api-reference#waituntil) have settled.
```

> If you want your users to be able to run `after`, you will have to provide you own implementation of `waitUntil` that behaves in an analogous way.

Reviewer: ```suggestion
If you want your users to be able to run `after`, you will have to provide your implementation of `waitUntil` that behaves in an analogous way.
```

Reviewer: Curious, what do we mean by analogous way?

### Interface

When `after` is called, Next.js will access `waitUntil` like this:

```jsx
const RequestContext = globalThis[Symbol.for('@next/request-context')]
const contextValue = RequestContext?.get()
const waitUntil = context?.waitUntil
```

Which means that `globalThis[Symbol.for('@next/request-context')]` is expected to contain an object like this:

```tsx
type NextRequestContext = {
  get(): NextRequestContextValue | undefined
}

type NextRequestContextValue = {
  waitUntil?: (promise: Promise<any>) => void
}
```

> ### Examples - providing `waitUntil` to Next.js

Reviewer: ```suggestion
### Examples
```

> Depending on your requirements, this can be implemented in multiple ways.

Reviewer: ```suggestion
Depending on your requirements, `waitUntil` can be provided to Next.js as **a wrapper around an `AsyncLocalStorage`** or as a **global object**:
```

> #### Example: A wrapper around an [`AsyncLocalStorage`](https://nodejs.org/api/async_context.html#class-asynclocalstorage)

Reviewer: Moving link outside hash to won't be clickable. Also adding more context to what the code is doing, but guessing here, does this make sense? 

```suggestion
#### A wrapper around an `AsyncLocalStorage`

You can use Node's [`AsyncLocalStorage`](https://nodejs.org/api/async_context.html#class-asynclocalstorage) API to store and manage the context for each incoming request. 
```

```tsx
import { AsyncLocalStorage } from 'node:async_hooks'

const RequestContextStorage = new AsyncLocalStorage<NextRequestContextValue>()

// Define and inject the accessor that next.js will use
const RequestContext: NextRequestContext = {
  get() {
    return RequestContextStorage.getStore()
  },
}
globalThis[Symbol.for('@next/request-context')] = RequestContext

const handler = (req, res) => {
  const contextValue = { waitUntil: YOUR_WAITUNTIL }
  // Provide the value
  return RequestContextStorage.run(contextValue, () => nextJsHandler(req, res))
}
```

> #### Example: A global object

Reviewer: ```suggestion
#### A global object
```

> Note that this may not be appropriate depending on how concurrency is handled on your platform.

Reviewer: Not sure how to word this, but trying to highlight why you'd go for a global object instead of a wrapper... 

```suggestion
You can define `waitUntil` directly as part of a global object, without the need to manage request-specific context. Although this approach is simpler, it's not appropriate... something about concurrency.

Note that this may not be appropriate depending on how concurrency is handled on your platform.
```
